svm/src/account_saver.rs

```diff
@@ -58,100 +51,87 @@ pub fn collect_accounts_to_store<'a>(
-        };
-
-        let message = tx.message();
-        let rollback_accounts = &loaded_transaction.rollback_accounts;
-        let maybe_nonce_address = rollback_accounts.nonce().map(|account| account.address());
-
-        for (i, (address, account)) in (0..message.account_keys().len())
-            .zip(loaded_transaction.accounts.iter_mut())
-            .filter(|(i, _)| is_storable_account(message, *i))
-        {
-            if message.is_writable(i) {
-                let should_collect_account = match execution_status {
-                    Ok(()) => true,
-                    Err(_) => {
-                        let is_fee_payer = i == 0;
-                        let is_nonce_account = Some(&*address) == maybe_nonce_address;
-                        post_process_failed_tx(
-                            account,
-                            is_fee_payer,
-                            is_nonce_account,
-                            rollback_accounts,
-                            durable_nonce,
-                            lamports_per_signature,
-                        );
-
-                        is_fee_payer || is_nonce_account
-                    }
-                };
+        if executed_tx.execution_details.status.is_ok() {
+            collect_accounts_for_successful_tx(
+                &mut accounts,
+                &mut transactions,
+                tx,
+                &executed_tx.loaded_transaction.accounts,
+            );
+        } else {
+            collect_accounts_for_failed_tx(
+                &mut accounts,
+                &mut transactions,
+                tx,
+                &mut executed_tx.loaded_transaction.rollback_accounts,
+                durable_nonce,
+                lamports_per_signature,
+            );
+        }
+    }
+    (accounts, transactions)
+}
 
-                if should_collect_account {
-                    // Add to the accounts to store
-                    accounts.push((&*address, &*account));
-                    transactions.push(Some(tx));
-                }
+fn collect_accounts_for_successful_tx<'a>(
+    collected_accounts: &mut Vec<(&'a Pubkey, &'a AccountSharedData)>,
+    collected_account_transactions: &mut Vec<Option<&'a SanitizedTransaction>>,
+    transaction: &'a SanitizedTransaction,
+    transaction_accounts: &'a [TransactionAccount],
+) {
+    let message = transaction.message();
+    for (i, (address, account)) in (0..message.account_keys().len()).zip(transaction_accounts) {
+        if message.is_writable(i) {
+            // Accounts that are invoked and also not passed as an instruction
+            // account to a program don't need to be stored because it's assumed
+            // to be impossible for a committable transaction to modify an
+            // invoked account if said account isn't passed to some program.
+            if !message.is_invoked(i) || message.is_instruction_account(i) {
```

Comment:
These checks are now inlined from the previously used `is_storable_account` check and they are called after `is_writable` because they are relatively slow